File: src/domes.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { initial, domeBounce } from "./game.js";
import { velocityUnit } from "./attributes.js";
const unit = 22.4 / 576;
test("both original bumper centers relaunch in eight directions with strict half-axis boundaries", () => {
  for (const center of [-256, 256])
    for (const [dx, dz, fx, fz] of [
      [12, 0, 1, 0],
      [-12, 0, -1, 0],
      [0, 12, 0, 1],
      [0, -12, 0, -1],
      [10, 10, 1, 1],
      [-10, 10, -1, 1],
      [10, -10, 1, -1],
      [-10, -10, -1, -1],
      [12, 6, 1, 0],
      [12, 7, 1, 1],
      [6, 12, 0, 1],
      [7, 12, 1, 1],
    ]) {
      const s = initial();
      Object.assign(s.ball, {
        owner: -1,
        lastTouch: 7,
        x: (center + dx) * unit,
        z: dz * unit,
        h: 0.25,
        vx: 1,
        vz: 2,
      });
      s.players[7].stats[4] = 200;
      domeBounce(s);
      assert.equal(s.ball.vx, fx * 8 * velocityUnit);
      assert.equal(s.ball.vz, fz * 8 * velocityUnit);
      assert.equal(s.ball.dirX, fx);
      assert.equal(s.ball.dirZ, fz);
      assert.equal(s.ball.speedTimer, 100);
      assert.equal(s.score[0], 2);
      assert.equal(
        s.ball.x,
        (center + dx) * unit,
        "contact does not push ball outside the bumper",
      );
    }
});
test("bumper excludes held/high balls, exact center and points outside original radius", () => {
  for (const change of [
    { owner: 7 },
    { h: 2 },
    { flightKind: 2, flightStage: 3 },
    { x: 256 * unit },
    { x: 272 * unit, z: 16 * unit },
    { x: 273 * unit },
  ]) {
    const s = initial();
    Object.assign(
      s.ball,
      { owner: -1, lastTouch: 7, x: 268 * unit, z: 0, h: 0.25, vx: -1 },
      change,
    );
    domeBounce(s);
    assert.equal(s.score[0], 0);
  }
  const s = initial();
  Object.assign(s.ball, {
    owner: -1,
    lastTouch: 7,
    x: 272 * unit,
    z: 0,
    h: 1.25,
    flightKind: 2,
    flightStage: 2,
  });
  s.multiplier = 2;
  domeBounce(s);
  assert.equal(s.score[0], 4);
  assert.equal(s.ball.flightIndex, -1);
  assert.equal(s.ball.flightStage, 1);
});

File: server/domes_test.go
package main

import "testing"

func TestBumperDirectionThresholdsAndSustain(t *testing.T) {
	for _, center := range []int{-256, 256} {
		for _, v := range [][4]int{{12, 0, 1, 0}, {-12, 0, -1, 0}, {0, 12, 0, 1}, {0, -12, 0, -1}, {10, 10, 1, 1}, {-10, 10, -1, 1}, {10, -10, 1, -1}, {-10, -10, -1, -1}, {12, 6, 1, 0}, {12, 7, 1, 1}, {6, 12, 0, 1}, {7, 12, 1, 1}} {
			s := initial()
			s.Ball = Ball{Owner: -1, LastTouch: 7, X: float64(center+v[0]) * terrainUnit, Z: float64(v[1]) * terrainUnit, H: .25, VX: 1, VZ: 2}
			s.Players[7].Stats[4] = 200
			s.domeBounce()
			if s.Ball.VX != float64(v[2])*8*velocityUnit || s.Ball.VZ != float64(v[3])*8*velocityUnit || s.Score[0] != 2 || s.Ball.SpeedTimer != 100 {
				t.Fatal("direction/score/sustain", center, v, s.Ball)
			}
		}
	}
}
func TestBumperEligibility(t *testing.T) {
	for _, change := range []func(*Ball){func(b *Ball) { b.Owner = 7 }, func(b *Ball) { b.H = 2 }, func(b *Ball) { b.FlightKind = 2; b.FlightStage = 3 }, func(b *Ball) { b.X = 256 * terrainUnit }, func(b *Ball) { b.X = 272 * terrainUnit; b.Z = 16 * terrainUnit }, func(b *Ball) { b.X = 273 * terrainUnit }} {
		s := initial()
		s.Ball = Ball{Owner: -1, LastTouch: 7, X: 268 * terrainUnit, H: .25, VX: -1}
		change(&s.Ball)
		s.domeBounce()
		if s.Score[0] != 0 {
			t.Fatal("ineligible bumper contact scored")
		}
	}
}
